### ISSUE TITLE
Chapel 1.18.0 rc0 Cray-XC module for aarch64 installs non-functional chpldoc

### ISSUE THREAD

**Issue by awallace-cray** [chapel-lang/chapel#11119]
### Summary of Problem
The full path to `python` (baked-in to chpldoc) points to a missing file, so sphinx, etc cannot run. 
However, `start_test` works just fine.

```
# 2018/09/13-14:09:46 awallace@kay-aarch:/cray/css/pe/dropoff
module load chapel

# 2018/09/13-14:10:21 awallace@kay-aarch:/cray/css/pe/dropoff
module swap PrgEnv-cray PrgEnv-gnu

# 2018/09/13-14:26:16 awallace@kay-aarch:/lus/scratch/awallace/build_configs5/test
chpldoc -o t release/examples/hello6-taskpar-dist.chpl
/opt/chapel/1.18.0/cray-xc/third-party/chpl-venv/install/cray-xc/py2.7/chpl-virtualenv/bin/sphinx-build: line 2: /lus/scratch/chapelu/jenkins/chapel-ci/kay-aarch/workspace/cray-module-release-arm-xc/chapel-1.18.0-cray-xc/chapel/third-party/chpl-venv/install/cray-xc/py2.7/chpl-virtualenv/bin/python: No such file or directory
error: building html output from chpldoc sphinx project
```

### Configuration Information

- Output of `chpl --version`:
```
# 2018/09/13-14:28:09 awallace@kay-aarch:/lus/scratch/awallace/build_configs5/test
type chpldoc
chpldoc is hashed (/opt/chapel/1.18.0/cray-xc/bin/cray-xc/chpldoc)

# 2018/09/13-14:34:44 awallace@kay-aarch:/lus/scratch/awallace/build_configs5/test
chpldoc --version
chpldoc version 1.18.0
Copyright (c) 2004-2018, Cray Inc.  (See LICENSE file for more details)
```
- Output of `$CHPL_HOME/util/printchplenv --anonymize`:
```
# 2018/09/13-14:38:11 awallace@kay-aarch:/lus/scratch/awallace/build_configs5/test
$CHPL_HOME/util/printchplenv --anonymize
CHPL_TARGET_PLATFORM: cray-xc
CHPL_TARGET_COMPILER: cray-prgenv-gnu
CHPL_TARGET_ARCH: arm-thunderx2
CHPL_LOCALE_MODEL: flat
CHPL_COMM: ugni
CHPL_TASKS: qthreads
CHPL_LAUNCHER: aprun
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: intrinsics
  CHPL_NETWORK_ATOMICS: ugni
CHPL_GMP: gmp
CHPL_HWLOC: hwloc
CHPL_REGEXP: re2
CHPL_AUX_FILESYS: none
```
- Back-end compiler and version, e.g. `gcc --version` or `clang --version`:
```
# 2018/09/13-14:37:34 awallace@kay-aarch:/lus/scratch/awallace/build_configs5/test
gcc --version
gcc (GCC) 7.3.0 20180125 (Cray Inc.)
```
- (For Cray systems only) Output of `module list`:
```
# 2018/09/13-14:35:16 awallace@kay-aarch:/lus/scratch/awallace/build_configs5/test
module list -t
Currently Loaded Modulefiles:
modules/3.2.10.6
alps/6.6.43-6.0.7.0_26.3__ga796da3.ari
nodestat/2.3.85-6.0.7.0_32.1__gc6218bb.ari
sdb/3.3.775-6.0.7.0_32.2__gb339c00.ari
udreg/2.3.2-6.0.7.0_33.11__g5196236.ari
ugni/6.0.14.0-6.0.7.0_23.1__gea11d3d.ari
gni-headers/5.0.12.0-6.0.7.0_24.1__g3b1768f.ari
dmapp/7.1.1-6.0.7.0_34.3__g5a674e0.ari
xpmem/2.2.15-6.0.7.1_5.7__g7549d06.ari
llm/21.3.530-6.0.7.0_39.1__g3b4230e.ari
nodehealth/5.6.13-6.0.7.0_61.2__ge9c9532.ari
system-config/3.5.2786-6.0.7.0_43.1__gc54785a.ari
Base-opts/2.4.135-6.0.7.0_38.1__g718f891.ari
gcc/7.3.0
craype-arm-thunderx2
craype-hugepages16M
craype-network-aries
craype/2.5.16.13
cray-mpich/7.7.3
rca/2.2.18-6.0.7.0_33.2__g2aa4f39.ari
chapel/1.18.0
cray-libsci/18.07.1
pmi/5.0.14
atp/2.1.3
perftools-base/7.0.4
PrgEnv-gnu/6.0.5
```

**Comment by awallace-cray** [comment]
2018-09-13 : The possibility exists that IF the Jenkins build workspace was visible in the environment where I ran these simple sanity-check tests on the system-installed Chapel release candidate Cray modules today, then ANOTHER incorrect baked-in path pointing back to the Jenkins build workspace might not have been caught by these tests. 
I think I have eliminated that possibility now, by verifying that the Jenkins build workspace was NOT visible; or if it was, by chmod'ing some intermediate directory to 700 and re-testing.

**Comment by awallace-cray** [comment]
Sure enough, file `/opt/chapel/1.18.0/cray-xc/third-party/chpl-venv/install/cray-xc/py2.7/chpl-virtualenv/bin/sphinx-build` is:
```
#!/bin/sh
'''exec' /lus/scratch/chapelu/jenkins/chapel-ci/kay-aarch/workspace/cray-module-release-arm-xc/chapel-1.18.0-cray-xc/chapel/third-party/chpl-venv/install/cray-xc/py2.7/chpl-virtualenv/bin/python "$0" "$@"
' '''
# -*- coding: utf-8 -*-
import re
import sys

from sphinx import main

if __name__ == '__main__':
    sys.argv[0] = re.sub(r'(-script\.pyw?|\.exe)?$', '', sys.argv[0])
    sys.exit(main())
```

**Comment by awallace-cray** [comment]
2018-09-13 14:50
The above problem seems similar to the problem this tries to solve:
- `third-party/chpl-venv/venv-use-sys-python.py` 
- only used by `make -C third-party/chpl-venv/use-system-python` 

Problem seems similar; but not identical. Ie, this does not look like it can be fixed just by adding that make command somewhere in the Cray aarch64 build script.

**Comment by awallace-cray** [comment]
2018-09-14 06:55
According to Ben A, 

> the hard-coded paths to python installations should be removed with the `virtualenv --relocatable` command. Perhaps that command is not working properly on Scout (or wherever the RPM is built).

**Comment by ben-albrecht** [comment]
> Some combination of pip, virtualenv, or sphinx version differences is likely responsible for the bad path in sphinx-build not being rewritten by `virtualenv --relocatable`

It would probably be helpful to understand the exact cause of `virtualenv --relocatable` failing if possible. The statement above sounds like you have an idea, but haven't yet confirmed.

**Comment by awallace-cray** [comment]
It is failing because the sphinx scripts installed in the aarch64 build use this shell exec thing to select which python binary to run; instead of the usual shebang line that `virtualenv --relocatable` knows how to fix, which the sphinx scripts installed in the x86_64 build use.  

Sphinx itself is the same version (1.3.3); that is controlled by some requirements file somewhere.

What makes the same version of sphinx end up with different version scripts? Whatever it is, it seems to happen before virtualenv.

**Comment by awallace-cray** [comment]
Based on saved log files from Jenkins release builds, the "Rewriting shebang" message stopped appearing in the Cray XC (x86_64) build logs sometime between April 19 and April 25, 2018.

**Comment by awallace-cray** [comment]
2018-09-16 - Tried using the same pre-installed copy of pip for the aarch64 module build as we have to use for the x86_64 build. Got a different version of setuptools installed, but have the same problem with chpldoc.

**Comment by awallace-cray** [comment]
2018-09-17 16:30 - Still no joy. 
Trying to just hack that sphinx-build script to see if I can get anything to work. 
This is the error that I cannot get past
```
# 2018/09/17-17:54:41 awallace@kay-aarch:/lus/scratch/awallace/build_configs5
env | grep CHPL
CHPL_MODULE_HOME=/cray/css/users/chapelu/jenkins/common.bogus/f920f80d2f2f611d2922c930f39fdd7b/chapel-test-opt/chapel/1.18.0.20180917/cray-xc
CHPL_HOST_PLATFORM=cray-xc
CHPL_CG_CPP_LINES=1
CHPL_TEST_NOMAIL=1
CHPL_HOME=/cray/css/users/chapelu/jenkins/common.bogus/f920f80d2f2f611d2922c930f39fdd7b/chapel-test-opt/chapel/1.18.0.20180917/cray-xc
# 2018/09/17-18:18:42 awallace@kay-aarch:/lus/scratch/awallace/build_configs5
chpldoc -o /tmp test/release/examples/hello6-taskpar-dist.chpl
Running Sphinx v1.3.3

Extension error:
Could not import extension sphinxcontrib.chapeldomain (exception: No module named sphinxcontrib.chapeldomain)
error: building html output from chpldoc sphinx project
```

This is the hack that helps it get that far. 
The "exec /usr/bin/python" came from the modded use-sys-python.py. The "export PYTHONPATH=..." came from me hacking things into that script by hand and trying to run chpldoc.
```
#!/bin/sh
'''export' PYTHONPATH=$CHPL_MODULE_HOME/third-party/chpl-venv/install/cray-xc/py2.7/chpl-virtualenv/lib/python2.7/site-packages:$PYTHONPATH; exec /usr/bin/python "$0" "$@"
' '''
# -*- coding: utf-8 -*-
import re
import sys

from sphinx import main

if __name__ == '__main__':
    sys.argv[0] = re.sub(r'(-script\.pyw?|\.exe)?$', '', sys.argv[0])
    sys.exit(main())
```

Subdir sphinxcontrib is right alongside sphinx, under site-packages. That PYTHONPATH should find both, you would think.

**Comment by awallace-cray** [comment]
2018-09-18 07:30 - There is hope. I started over from scratch with the new build scripts (https://github.com/awallace-cray/chapel/tree/build_configs5), logged in with awallace not chapelu, no Jenkins. No special python 2.7 version, but through an oversight, I did have the env variables CHPL_PIP, CHPL_PYTHONPATH, etc from the x86_64 Cray module Jenkins build. 
Built Chapel in one configuration only, but with chpldoc, test-venv, and mason. (Whoa- check the order those were built in?)  
chpldoc works. The sphinx-build script does not have the flawed "exec" shebang line. It has the regular shebang line which virtualenv successfully converts to "relative" form as shown in the log.
